### PR TITLE
[Deprecated] enable dynamo ci on Windows.

### PR DESCRIPTION
**Pull request by xuhancn**
I'm working with @chuanqi129 to enable inductor UTs on Windows. Early we try to run the command:
```cmd
python test/run_test.py --exclude-jit-executor --exclude-distributed-tests --verbose
```
It is strange behavior:
1. It works well on Linux.
2. But it was quit on Windows silently, without any message prompted.

After some debug work, we found `torch/_dynamo/test_case.py`, `run_tests` function will skip all UTs on Windows, without any message. It seems a global switch and it turns off all dynamo related UT.

So, I'm turned on the global switch for dynamo on Windows. 
After that, all dynamo/inductor UTs are triggered, and shows a lot of failures. And then, I take a lot of effort to skip the UTs, what failed on Windows.

**TODO:**
We still need review and enabled these skipped UTs in the further.

cc @peterjc123 @mszhanyi @skyline75489 @nbcsm @iremyux @Blackhex @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Comment by xuhancn** [comment]
Still need to skip more UTs for `ciflow/periodic`'s `cuda` on Windows.

**Comment by xuhancn** [comment]
image

Hi @malfet, @jansel, @desertfire, @ezyang 

Good news is the inductor/dynamo UTs is enabled by this PR. But the test node 3, runs too long time to pass test.
How to balance some workload to other nodes?

File: .github/workflows/pull.yml [hunk 1]
```diff
@@ -391,9 +391,10 @@ jobs:
       runner: "${{ needs.get-label-type.outputs.label-type }}windows.4xlarge.nonephemeral"
       test-matrix: |
         { include: [
-          { config: "default", shard: 1, num_shards: 3, runner: "${{ needs.get-label-type.outputs.label-type }}windows.4xlarge.nonephemeral" },
-          { config: "default", shard: 2, num_shards: 3, runner: "${{ needs.get-label-type.outputs.label-type }}windows.4xlarge.nonephemeral" },
-          { config: "default", shard: 3, num_shards: 3, runner: "${{ needs.get-label-type.outputs.label-type }}windows.4xlarge.nonephemeral" },
+          { config: "default", shard: 1, num_shards: 4, runner: "${{ needs.get-label-type.outputs.label-type }}windows.4xlarge.nonephemeral" },
+          { config: "default", shard: 2, num_shards: 4, runner: "${{ needs.get-label-type.outputs.label-type }}windows.4xlarge.nonephemeral" },
+          { config: "default", shard: 3, num_shards: 4, runner: "${{ needs.get-label-type.outputs.label-type }}windows.4xlarge.nonephemeral" },
+          { config: "default", shard: 4, num_shards: 4, runner: "${{ needs.get-label-type.outputs.label-type }}windows.4xlarge.nonephemeral" },
```
Comment (jgong5) [inline .github/workflows/pull.yml:397]: Why do we need to add shard "4"?

File: .github/workflows/trunk.yml [hunk 2]
```diff
@@ -195,9 +195,10 @@ jobs:
       runner: "${{ needs.get-label-type.outputs.label-type }}windows.4xlarge.nonephemeral"
       test-matrix: |
         { include: [
-          { config: "default", shard: 1, num_shards: 3, runner: "${{ needs.get-label-type.outputs.label-type }}windows.4xlarge.nonephemeral" },
-          { config: "default", shard: 2, num_shards: 3, runner: "${{ needs.get-label-type.outputs.label-type }}windows.4xlarge.nonephemeral" },
-          { config: "default", shard: 3, num_shards: 3, runner: "${{ needs.get-label-type.outputs.label-type }}windows.4xlarge.nonephemeral" },
+          { config: "default", shard: 1, num_shards: 4, runner: "${{ needs.get-label-type.outputs.label-type }}windows.4xlarge.nonephemeral" },
+          { config: "default", shard: 2, num_shards: 4, runner: "${{ needs.get-label-type.outputs.label-type }}windows.4xlarge.nonephemeral" },
+          { config: "default", shard: 3, num_shards: 4, runner: "${{ needs.get-label-type.outputs.label-type }}windows.4xlarge.nonephemeral" },
+          { config: "default", shard: 4, num_shards: 4, runner: "${{ needs.get-label-type.outputs.label-type }}windows.4xlarge.nonephemeral" },
```
Comment (jgong5) [inline .github/workflows/trunk.yml:201]: ditto

**Review by jgong5 — APPROVED** [review]
LGTM. A good first step to enable Windows UTs for dynamo!

File: .github/workflows/pull.yml (hunk 1 above)
Comment (xuhancn) [reply .github/workflows/pull.yml:397]: https://github.com/pytorch/pytorch/pull/133226#issuecomment-2287700510 3 shared always timeout and fail. Add one node to balance load.

**Comment by ezyang** [comment]
> Good news is the inductor/dynamo UTs is enabled by this PR. But the test node 3, runs too long time to pass test.
How to balance some workload to other nodes?

Need to make some more shards. Dev infra should help

File: .github/workflows/pull.yml [hunk 3]
```diff
@@ -394,6 +394,7 @@ jobs:
           { config: "default", shard: 1, num_shards: 3, runner: "${{ needs.get-label-type.outputs.label-type }}windows.4xlarge.nonephemeral" },
           { config: "default", shard: 2, num_shards: 3, runner: "${{ needs.get-label-type.outputs.label-type }}windows.4xlarge.nonephemeral" },
           { config: "default", shard: 3, num_shards: 3, runner: "${{ needs.get-label-type.outputs.label-type }}windows.4xlarge.nonephemeral" },
+          { config: "inductor_cpp_wrapper_abi_compatible", shard: 1, num_shards: 1, runner: "${{ needs.get-label-type.outputs.label-type }}windows.4xlarge.nonephemeral" },
```
Comment (ZainRizvi) [inline .github/workflows/pull.yml:397]: Might the inductor.yml workflow be a more appropriate home for this job?

**Review by jansel — CHANGES_REQUESTED** [review]
1) This is adding a lot of `skipIfWindows` in places I would expect to work fine on Windows, you can provide a summary of the main underlying reasons for the skips?
2) Let's increase the number of shards, so tests stay <4 hours

**Comment by xuhancn** [comment]
> 1. This is adding a lot of `skipIfWindows` in places I would expect to work fine on Windows, you can provide a summary of the main underlying reasons for the skips?
> 2. Let's increase the number of shards, so tests stay <4 hours

Hi @jansel 
For `skipIfWindows`, I known many expected work UT are passed. It is temporary. I want to enable CI firstly and it will help avoid regression. 
Still some skiped UTs are WIP, I write a small script to trace them. I will fix them and remove `skipIfWindows` later.
Current Skiped summary:
```cmd
(win_mkl_static) D:\xu_git\dnnl_cb\pytorch>python count_skip.py
Skip UTs:
D:\xu_git\dnnl_cb\pytorch\test\dynamo\test_activation_checkpointing.py: @skipIfWindows num: 1
D:\xu_git\dnnl_cb\pytorch\test\dynamo\test_after_aot.py: @skipIfWindows num: 1
D:\xu_git\dnnl_cb\pytorch\test\dynamo\test_aot_autograd.py: @skipIfWindows num: 2
D:\xu_git\dnnl_cb\pytorch\test\dynamo\test_aot_autograd_cache.py: @skipIfWindows num: 8
D:\xu_git\dnnl_cb\pytorch\test\dynamo\test_autograd_function.py: @skipIfWindows num: 3
D:\xu_git\dnnl_cb\pytorch\test\dynamo\test_backward_higher_order_ops.py: @skipIfWindows num: 4
D:\xu_git\dnnl_cb\pytorch\test\dynamo\test_comptime.py: @skipIfWindows num: 6
D:\xu_git\dnnl_cb\pytorch\test\dynamo\test_exc.py: @skipIfWindows num: 8
D:\xu_git\dnnl_cb\pytorch\test\dynamo\test_export.py: @skipIfWindows num: 34
D:\xu_git\dnnl_cb\pytorch\test\dynamo\test_functions.py: @skipIfWindows num: 20
D:\xu_git\dnnl_cb\pytorch\test\dynamo\test_hooks.py: @skipIfWindows num: 12
D:\xu_git\dnnl_cb\pytorch\test\dynamo\test_input_attr_tracking.py: @skipIfWindows num: 3
D:\xu_git\dnnl_cb\pytorch\test\dynamo\test_interop.py: @skipIfWindows num: 3
D:\xu_git\dnnl_cb\pytorch\test\dynamo\test_logging.py: @skipIfWindows num: 9
D:\xu_git\dnnl_cb\pytorch\test\dynamo\test_minifier.py: @skipIfWindows num: 5
D:\xu_git\dnnl_cb\pytorch\test\dynamo\test_misc.py: @skipIfWindows num: 45
D:\xu_git\dnnl_cb\pytorch\test\dynamo\test_modules.py: @skipIfWindows num: 2
D:\xu_git\dnnl_cb\pytorch\test\dynamo\test_repros.py: @skipIfWindows num: 18
D:\xu_git\dnnl_cb\pytorch\test\dynamo\test_structured_trace.py: @skipIfWindows num: 9
D:\xu_git\dnnl_cb\pytorch\test\dynamo\test_subclasses.py: @skipIfWindows num: 8
D:\xu_git\dnnl_cb\pytorch\test\dynamo\test_trace_rules.py: @skipIfWindows num: 1
D:\xu_git\dnnl_cb\pytorch\test\dynamo\test_unspec.py: @skipIfWindows num: 7
D:\xu_git\dnnl_cb\pytorch\test\inductor\test_codecache.py: @skipIfWindows num: 8
D:\xu_git\dnnl_cb\pytorch\test\inductor\test_compile_worker.py: @skipIfWindows num: 3
D:\xu_git\dnnl_cb\pytorch\test\inductor\test_config.py: @skipIfWindows num: 1
D:\xu_git\dnnl_cb\pytorch\test\inductor\test_cpu_repro.py: @skipIfWindows num: 20
D:\xu_git\dnnl_cb\pytorch\test\inductor\test_debug_trace.py: @skipIfWindows num: 1
D:\xu_git\dnnl_cb\pytorch\test\inductor\test_distributed_patterns.py: @skipIfWindows num: 5
D:\xu_git\dnnl_cb\pytorch\test\inductor\test_indexing.py: @skipIfWindows num: 5
D:\xu_git\dnnl_cb\pytorch\test\inductor\test_inductor_utils.py: @skipIfWindows num: 2
D:\xu_git\dnnl_cb\pytorch\test\inductor\test_minifier.py: @skipIfWindows num: 5
D:\xu_git\dnnl_cb\pytorch\test\inductor\test_minifier_isolate.py: @skipIfWindows num: 1
D:\xu_git\dnnl_cb\pytorch\test\inductor\test_torchinductor.py: @skipIfWindows num: 84
D:\xu_git\dnnl_cb\pytorch\test\export\test_experimental.py: @skipIfWindows num: 2
D:\xu_git\dnnl_cb\pytorch\test\export\test_export.py: @skipIfWindows num: 3
Skip Blocks:
D:\xu_git\dnnl_cb\pytorch\test\dynamo\test_dynamic_shapes.py: if not IS_WINDOWS: num: 2
D:\xu_git\dnnl_cb\pytorch\test\dynamo\test_logging.py: if not IS_WINDOWS: num: 2
D:\xu_git\dnnl_cb\pytorch\test\inductor\test_mmdecomp.py: if not IS_WINDOWS: num: 1
Skip whole file:
D:\xu_git\dnnl_cb\pytorch\test\dynamo\test_higher_order_ops.py: if IS_WINDOWS and IS_CI: num: 1
D:\xu_git\dnnl_cb\pytorch\test\inductor\test_benchmark_fusion.py: if IS_WINDOWS and IS_CI: num: 1
D:\xu_git\dnnl_cb\pytorch\test\inductor\test_binary_folding.py: if IS_WINDOWS and IS_CI: num: 1
D:\xu_git\dnnl_cb\pytorch\test\inductor\test_compiled_autograd.py: if IS_WINDOWS and IS_CI: num: 1
D:\xu_git\dnnl_cb\pytorch\test\inductor\test_compiled_optimizers.py: if IS_WINDOWS and IS_CI: num: 1
D:\xu_git\dnnl_cb\pytorch\test\inductor\test_cudagraph_trees.py: if IS_WINDOWS and IS_CI: num: 1
D:\xu_git\dnnl_cb\pytorch\test\inductor\test_efficient_conv_bn_eval.py: if IS_WINDOWS and IS_CI: num: 1
D:\xu_git\dnnl_cb\pytorch\test\inductor\test_halide.py: if IS_WINDOWS and IS_CI: num: 1
D:\xu_git\dnnl_cb\pytorch\test\inductor\test_inductor_freezing.py: if IS_WINDOWS and IS_CI: num: 1
D:\xu_git\dnnl_cb\pytorch\test\inductor\test_memory_planning.py: if IS_WINDOWS and IS_CI: num: 1
D:\xu_git\dnnl_cb\pytorch\test\inductor\test_torchinductor_codegen_dynamic_shapes.py: if IS_WINDOWS and IS_CI: num: 1
D:\xu_git\dnnl_cb\pytorch\test\inductor\test_torchinductor_dynamic_shapes.py: if IS_WINDOWS and IS_CI: num: 1
D:\xu_git\dnnl_cb\pytorch\test\inductor\test_xpu_basic.py: if IS_WINDOWS and IS_CI: num: 1
```
The count script:
```python
import os

def enum_files(root_dir: str):
    module_files = []
    for root, dirs, files in os.walk(root_dir):
        for file in files:
            if file.endswith(".py"):
                module_files.append(os.path.join(root, file))
    return module_files


def count_num(file_path, str_to_count):
    file=open(file_path)
    try:
        ctx = file.read()
        count_num = ctx.count(str_to_count)
    except:
        return 0
    return count_num


def count_str_in_dir(dir_path, str_to_count):
    files = enum_files(dir_path)
    for i in files:
        num = count_num(i, str_to_count)
        if num != 0:
            print(f"{i}: {str_to_count} num: {num}")

def count_skip_windows_uts():
    cur_dir = os.getcwd()
    dynamo_dir = os.path.join(cur_dir, "test", "dynamo")
    inductor_dir = os.path.join(cur_dir, "test", "inductor")
    export_dir = os.path.join(cur_dir, "test", "export")

    print("Skip UTs:")
    count_str_in_dir(dynamo_dir, "@skipIfWindows")
    count_str_in_dir(inductor_dir, "@skipIfWindows")
    count_str_in_dir(export_dir, "@skipIfWindows")
    
    print("Skip Blocks:")
    count_str_in_dir(dynamo_dir, "if not IS_WINDOWS:")
    count_str_in_dir(inductor_dir, "if not IS_WINDOWS:")
    count_str_in_dir(export_dir, "if not IS_WINDOWS:")

    print("Skip whole file:")
    count_str_in_dir(dynamo_dir, "if IS_WINDOWS and IS_CI:")
    count_str_in_dir(inductor_dir, "if IS_WINDOWS and IS_CI:")
    count_str_in_dir(export_dir, "if IS_WINDOWS and IS_CI:")

if __name__ == "__main__":
    count_skip_windows_uts()
```

For reduce CI time, @chuanqi129 will help on research how to optimize ciflow.

CC: @ezyang , @ZainRizvi

**Comment by chuanqi129** [comment]
Hi @xuhancn @ezyang @ZainRizvi , I have checked the shard logic of the pytorch UT test. The UT test cases shard based on test class in the test files, refer https://github.com/pytorch/pytorch/blob/main/tools/testing/test_selections.py#L185. And those 2 test files `test/inductor/test_torchinductor.py` and `test/inductor/test_cpu_cpp_wrapper.py` just have one test case class [CpuTests](https://github.com/pytorch/pytorch/blob/main/test/inductor/test_torchinductor.py#L11356), it means those 2 test files can't be sharded for cpu only test even we already used shard_num=4. We may need to refactor those 2 test files with multiple test case classes.

**Comment by xuhancn** [comment]
> Hi @xuhancn @ezyang @ZainRizvi , I have checked the shard logic of the pytorch UT test. The UT test cases shard based on test class in the test files, refer https://github.com/pytorch/pytorch/blob/main/tools/testing/test_selections.py#L185. And those 2 test files `test/inductor/test_torchinductor.py` and `test/inductor/test_cpu_cpp_wrapper.py` just have one test case class [CpuTests](https://github.com/pytorch/pytorch/blob/main/test/inductor/test_torchinductor.py#L11356), it means those 2 test files can't be sharded for cpu only test even we already used shard_num=4. We may need to refactor those 2 test files with multiple test case classes.

Thank you @chuanqi129 , in my dev PC (I9 11900KB). each of `test/inductor/test_torchinductor.py` and `test/inductor/test_cpu_cpp_wrapper.py` should take about 2 hours. We need to split them.

**Comment by xuhancn** [comment]
> Hi @xuhancn @ezyang @ZainRizvi , I have checked the shard logic of the pytorch UT test. The UT test cases shard based on test class in the test files, refer https://github.com/pytorch/pytorch/blob/main/tools/testing/test_selections.py#L185. And those 2 test files `test/inductor/test_torchinductor.py` and `test/inductor/test_cpu_cpp_wrapper.py` just have one test case class [CpuTests](https://github.com/pytorch/pytorch/blob/main/test/inductor/test_torchinductor.py#L11356), it means those 2 test files can't be sharded for cpu only test even we already used shard_num=4. We may need to refactor those 2 test files with multiple test case classes.

@atalman any comments about this?

**Comment by jansel** [comment]
My question was more about *why* the tests were failing, not how many are failing.  What is the most common error message?

**Comment by xuhancn** [comment]
> My question was more about _why_ the tests were failing, not how many are failing. What is the most common error message?

Hi @jansel 
1. After some other PRs merged, the failed PR will decresment a lot. 
2. Common message if from here: https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_utils.py#L1813

**Comment by xuhancn** [comment]
> > My question was more about _why_ the tests were failing, not how many are failing. What is the most common error message?
> 
> Hi @jansel
> 
> 1. After some other PRs merged, the failed PR will decresment a lot.
> 2. Common message if from here: https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_utils.py#L1813

Hi @jansel , I will calibration UTs after some key PR landed. So many UTs fixed now: https://github.com/pytorch/pytorch/pull/134033

**Comment by jansel** [comment]
Sounds good, filling the messages explaining the skips would be very helpful as well.

**Pull request closed without merging**